Create test to ensure we don't unusually increase size of BSI

dylan-matthew-turner opened the pull request:
The size of our images will likely slowly grow over time, and potentially we could increase faster than we are prepared for. We need to make sure we don't exceed the storage capacity of our smaller devices anytime soon.

Using an estimation of how fast we can grow gives us a value that can be added to our current image size to set the max amount we're allowed to exceed without intervention through 2024. This value is different for ARM (~300MB) and x86_64 (~1400MB).

I've added this test to the existing base system image tests, which run on a newly installed base system image which we can get the install size of to compare against our threshold. The failure state is simply "does the install size exceed the max-size threshold."

This has been run on a cRIO-9037 to verify the test will function.

amstewart commented on recipes-ni/ni-base-system-image-tests/files/disk_size.py:
Seems like this commit should be squashed into the previous, since these commits' scripts call each other.

amstewart commented on recipes-ni/ni-base-system-image-tests/files/disk_size.py:
```diff
@@ -0,0 +1,89 @@
+X86_INC_RATE = (X86_SMALLEST_DRIVE - X86_BASE_SIZE) / (X86_EOL_YEAR - BASE_YEAR)
+X86_MAX_SIZE_EOY = (X86_INC_RATE * (NEXT_YEAR - BASE_YEAR)) + X86_BASE_SIZE
+
+class Logger:
+    prefix_logs = []
+    logs = []
+    def log(self, log):
+        self.logs.append(log)
+
+    def prefix_log(self, log):
+        self.prefix_logs.append(log)
+
+    def report(self):
+        for log in self.prefix_logs:
+            print(log)
+        for log in self.logs:
+            print(log)
```
Is there a reason you're not using the standard python `logging` module for this? That module is pretty neat and I use it all the time.

blueOkiris replied:
This is how our other tests do it, but outside of that, no

chaitu236 replied:
The only reason our other tests used this was because of prefix_log. It looks like you're not using it so you can probably just use logging module

amstewart commented on recipes-ni/ni-base-system-image-tests/files/disk_size.py:
```diff
@@ -0,0 +1,89 @@
+X86_SMALLEST_DRIVE = 4096   # The size of cRIO ___'s hardrive
+X86_EOL_YEAR = 2038         # When the cRIO ___ goes EOL
+X86_INC_RATE = (X86_SMALLEST_DRIVE - X86_BASE_SIZE) / (X86_EOL_YEAR - BASE_YEAR)
+X86_MAX_SIZE_EOY = (X86_INC_RATE * (NEXT_YEAR - BASE_YEAR)) + X86_BASE_SIZE
```
I know you replace this block later in the patchset, but could you squash down the changes do that there aren't false advertisements of EOL dates in the git history?

blueOkiris replied:
Good catch

amstewart commented on recipes-ni/ni-base-system-image-tests/files/run-ptest:
The commit message here should be `ni-base-system-image-tests: ...`.

But honestly, I think this could also be silently squashed into the previous commits.

amstewart commented on recipes-ni/ni-base-system-image-tests/ni-base-system-image-tests.bb:
Squash.

amstewart commented on recipes-ni/ni-base-system-image-tests/files/disk_size.py:
```diff
@@ -0,0 +1,89 @@
+def get_disk_used():
+    # Format: "####M      /"
+    root_size = int(run_cmd([ 'du', '-hs', '-BM', '/' ]).split('\t')[0].strip()[:-1])
+    boot_size = int(run_cmd([ 'du', '-hs', '-BM', '/boot' ]).split('\t')[0].strip()[:-1])
```
I think these operations are complicated enough that they should be moved into their own method. eg.

```python3
def du_path(path):
    proc = subprocess.run(["du", "-hs", ...] + [path], ...)
    result = proc.stdout ...
    return result
```

SparkingSpork commented on recipes-ni/ni-base-system-image-tests/files/disk_size.py:
```diff
@@ -0,0 +1,76 @@
+
+# See internal wiki for calculation
+X86_MAX_SIZE_EOY = 1400
+ARM_MAX_SIZE_EOY = 300
```
I'd suggest making these parameters for the script that the ptest bash script passes instead of constants in the script. We might have additional limits in the future (e.g., if we add ARM 64-bit support or other images) and having them defined in the test script might be easier to read or update.

dylan-matthew-turner committed:
ni-base-system-image-test: ctrl disk size inc rate

- Added disk_size.py, a test that checks installed size against a pre-determined max
  size for BSI
- Added test_disk_size.sh, which calls the disk_size.py in the way that
  ptest expects
- Modified run-ptest to call test_disk_size.sh
- Modified ni-base-system-image-tests.bb to install the new files

Signed-off-by: Dylan Turner <dylan.turner@ni.com>

dylan-matthew-turner commented:
Changes:

- Moved size constants to CLI arguments
- Squashed commits into one since all the scripts call each other
- Pulled out du command into its own function
- Switched to Python logging library instead
